### ISSUE TITLE
apply for pre-trained weight of ner model and ere model

### ISSUE THREAD

**Issue by xerrors** [yanzhh/HGERE#4]
Thank you very much for open-sourcing your highly readable code. I would like to request access to your pre-trained model for testing purposes. I would be extremely grateful if it's possible. Email: wenjie.zhang@stu.jiangnan.edu.cn

**Comment by xerrors** [comment]
Or can you provide the inference overhead of the model in the NER and RE phases? I'm getting poor results from training on my device and the inference time to generate the predictions file is long (90seconds for NER testing on ACE05). I'm not sure if this is normal.

**Comment by yanzhh** [comment]
Thank you for your interest in our project. Unfortunately, due to my graduation, the trained model weights are no longer available from the university server.
Regarding inference times, a Bert-base model handling NER and RE on ACE05 takes around 50+ seconds with an A40 GPU (48G) and roughly 80-90 seconds on a 2080Ti.
Regarding your concerns about poor results, based on previous training experiences, the optimal parameters can differ based on the devices. Therefore, you may need parameter tuning for best performance.


________________________________
发件人: Wenjie Zhang ***@***.***>
发送时间: 2024年5月13日 13:20
收件人: yanzhh/HGERE ***@***.***>
抄送: Subscribed ***@***.***>
主题: Re: [yanzhh/HGERE] apply for pre-trained weight of ner model and ere model (Issue #4)


Or can you provide the inference overhead of the model in the NER and RE phases? I'm getting poor results from training on my device and the inference time to generate the predictions file is long (90seconds for NER testing on ACE05). I'm not sure if this is normal.

―

**Comment by xerrors** [comment]
Thank you for your reply. I adjusted the inference strategy and finally got an inference time of 48s + 73s = 121s on the 3090, which is basically close to your result. Thank you again for your reply.

**Comment by yanzhh** [comment]
Our model is designed to perform joint inference for NER and RE tasks given the pruned entity spans, and it generally produces a single inference time. The inference time on ACE05 test dataset (bert-base) is less than 90s on a 2080Ti so it should be less than 90s on a 3090. Could you please clarify what the 48s and 73s in your measurement correspond to?

**Comment by xerrors** [comment]
On the ACE05 test set, it took 50.70s on run_pruner

```log
2024-05-13 13:50:57,380 [MainThread  ] [INFO ]  Evaluate the following checkpoints: ['saves/ace05ner_models/pruner/biafencoder_seq300-spanlen8-rank768-hid768-span256-entnum3-18-lr1e-5-extraattn/ace05-bert-42/checkpoint-9102']
2024-05-13 13:50:57,380 [MainThread  ] [INFO ]  loading weights file saves/ace05ner_models/pruner/biafencoder_seq300-spanlen8-rank768-hid768-span256-entnum3-18-lr1e-5-extraattn/ace05-bert-42/checkpoint-9102/pytorch_model.bin
2024-05-13 13:50:59,251 [MainThread  ] [INFO ]  Evaluate on test.json
2024-05-13 13:50:59,824 [MainThread  ] [INFO ]  maxL: 1382
2024-05-13 13:50:59,824 [MainThread  ] [INFO ]  maxR: 298
2024-05-13 13:50:59,826 [MainThread  ] [INFO ]  ***** Running evaluation 9102 *****
2024-05-13 13:50:59,826 [MainThread  ] [INFO ]    Num examples = 7640
2024-05-13 13:50:59,826 [MainThread  ] [INFO ]    Batch size = 8
Evaluating: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 955/955 [00:50<00:00, 18.84it/s]
2024-05-13 13:51:50,535 [MainThread  ] [INFO ]    Evaluation done in total 50.709055 secs (150.663428 example per second)
2024-05-13 13:51:50,535 [MainThread  ] [INFO ]  Result: {'f1': '0.5725', 'precision': '0.4015', 'recall': '0.9973', 'f1_overlap': '0.4335', 'p_overlap': '0.2769', 'r_overlap': '0.9973'}
2024-05-13 13:51:50,544 [MainThread  ] [INFO ]  evaluate test file into ent_pred_test.json done.
2024-05-13 13:51:50,554 [MainThread  ] [INFO ]  evaluate test file into ent_pred_test_overlap.json done.
```

69.97s on run_hgnn.

```log
2024-05-13 20:13:22,631 [MainThread  ] [INFO ]  Evaluate the following checkpoints: ['saves/HGERE/ace05re_models/bert/tersibcop/facencbiaf-seq256-mem400-iter3-eps1e-8-layernorm+-attnself/ent400-rel400-lr2e-5-5e-5-bs18-ep15/Hyper_ace05_bert-42/checkpoint-25290']
2024-05-13 20:13:22,631 [MainThread  ] [INFO ]  loading weights file saves/HGERE/ace05re_models/bert/tersibcop/facencbiaf-seq256-mem400-iter3-eps1e-8-layernorm+-attnself/ent400-rel400-lr2e-5-5e-5-bs18-ep15/Hyper_ace05_bert-42/checkpoint-25290/pytorch_model.bin
2024-05-13 20:13:23,717 [MainThread  ] [INFO ]  Evaluate on saves/ace05ner_models/pruner/biafencoder_seq300-spanlen8-rank768-hid768-span256-entnum3-18-lr1e-5_biafmode-extraattn/ace05-bert-46/ent_pred_test.json
2024-05-13 20:13:23,718 [MainThread  ] [INFO ]  ***** Running evaluation 25290 *****
2024-05-13 20:13:23,718 [MainThread  ] [INFO ]    Batch size = 32
2024-05-13 20:13:23,718 [MainThread  ] [INFO ]  loading from saves/ace05ner_models/pruner/biafencoder_seq300-spanlen8-rank768-hid768-span256-entnum3-18-lr1e-5_biafmode-extraattn/ace05-bert-46/ent_pred_test.json
2024-05-13 20:13:24,429 [MainThread  ] [INFO ]  maxR: 86
2024-05-13 20:13:24,429 [MainThread  ] [INFO ]  maxL: 1501
2024-05-13 20:13:24,429 [MainThread  ] [INFO ]    Num examples = 2050
Evaluating: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 542/542 [01:09<00:00,  7.77it/s]
2024-05-13 20:14:34,399 [MainThread  ] [INFO ]    Evaluation done in total 69.970087 secs (29.298234 example per second)
2024-05-13 20:14:34,400 [MainThread  ] [INFO ]  Result:ner_p:0.3465, ner_r:0.9354, ner_f1:0.5057; rel_p:0.4182, rel_r:0.6577, rel_f1:0.5113; rel_p+:0.4028, rel_r+:0.6334, rel_f1+:0.4924
```

I'm not sure if my understanding is accurate. pruner first predicts entities based on the sentences in the dataset, and then hgnn classifies relations based on the predicted entities.

So if I were to calculate the overall overhead, would I be summing the runtime of the two scripts? Is it possible for hgnn alone to predict triples directly on unstructured sentences?

**Comment by yanzhh** [comment]
First the pruner does not do the NER task, it only filter out the least likely entity spans and produces a candidate span set (so it has a low precision but very high recall), then the joint ERE model (hgnn) performs the NER and RE tasks jointly on this set. 
For all ERE experiments, we consistently work with the optimal candidate span set. Since we do not need to rerun the pruner for each ERE experiment, we generally do not include the runtime of pruner for the inference of  ERE model.
It seems your ERE results are quit unusual. While your recall is similar to mine, there is a significant difference in precision.

**Comment by xerrors** [comment]
Thanks for the detailed explanation. I would like to elaborate further on my doubts. If I apply the model to an actual production environment and face 10k unseen sentences, do I need to run pruner to get the best candidate span set before running ERE?

**Comment by yanzhh** [comment]
It seems there might be a slight misunderstanding regarding our framework. Let me  clarify how it works. Our framework tackles unstructured text in two main steps:
1. Firstly, a pruner is employed to produces a candidate span set,  efficiently filtering out the least likely entity spans while retaining almost all probable ones.
2. Then, the joint ERE model identifies whether each span in this set is indeed an entity (NER task) and classify the relation between each span pair of this set (RE task). 

Therefore, when dealing with a new dataset, it is essential to first run the pruner to generate the optimal candidate span set specific to that dataset. Then you can train our joint ERE model on this span set.
